libgedathon/docs/gsym-set-font-size.1.in
.TH gsym-set-font-size 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
gsym-set-font-size \- gEDA-gaf Symbol File Utility

.SH SYNOPSIS
.B gsym-set-font-size
.RI  [ INFORMATION ] | [ OPTIONS ] [ DIRECTIVES ] [INPUT] [OUTPUT]

.SH DESCRIPTION
.B gsym-set-font-size
Sets the font size of Text attributes in a symbol file to default values or as otherwise specified

.SH QUICK
.B usage: [-h] | [Options] -i <inputfile> [[-o] <outputfile> ]

Example: Set the font size of all attributes to default values in all symbol files in the current directory and in all subdirectories:

          gsym-set-font-size.py -r

.SH INFORMATION
Use of the following options displays information related to gsym-set-font-size
but does not inspect or alter any symbol files:
.IP "-h, --help"
Show information about the advanced usage of this command.
.IP "-V, --version"
Displays software version information.

.SH OPTIONS
The following options can be used to modify the default behavior of gsym-set-font-size
.TP
The following list of attribute flags correspond to attributes set to 10 points by default, if these options are used the corresponding attributes will not be modified:

.IP "-D, --device"
.IP "-R, --refdes"
.IP "-A, --author"
.IP "-U, --numslots"
.IP "-E, --description"
.IP "-C, --comment"
.IP "-N, --net"
.IP "-M, --netname"
.IP "-V, --value"
.IP "-L, --pinlabel"

.TP
The following list of attribute flags correspond to attributes set to 8 points by default, if these options are used the corresponding attributes will not be modified:

.IP "-q, --pinseq"
.IP "-p, --pinnumber"
.IP "-t, --pintype"

.TP
The following list of attributes are set to 10 points, and no option is provided to change this behavior:

"documentation", "symversion", "dist-license", "use-license", "file", "footprint", "footprints",
"graphical", "pins", "slot", "slotdef", "spice-type", "electtype", "mechtype", "model-name"

.SH DIRECTIVES
.IP "-v, --verbose"
Enable verbose mode. Displays additional information about what the program is doing.
.IP "-r, --recursive"
Process all symbol files in the current and all subordinated directories.
.IP "-i, --input <input file>"
Specify the name of the input sym file. Wild cards are accepted, the default is *.sym.
.IP "-o, --output <output file>"
Specifies the name of the output sym file. When the output file name is not specified the input file will be over written

.SH HOMEPAGE
http://www.geda-project.org/

.SH BUGS
You can report bugs at the Launchpad site of gsym-set-font-size:
https://bugs.launchpad.net/geda/gsym-set-font-size/+filebug

.SH AUTHOR
Wiley Edward Hill < wileyhill dot gmail dot com >

.SH SEE ALSO
.BR gsym-attribute (1) gsym-set-visibility (1) geda-snap-grid (1)

.SH COPYRIGHT
.nf
Copyright \(co @YEAR@ Wiley Edward Hill.  License GPLv3+: GNU GPL
version 3.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2017 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf gEDA 2.1.1.1 - 11/12/17 Rampage Edition

* Expanded support for internationalization (i18n).

* The build system now support using the --with-included-gettext option to
  configure the suite to build and use an in-source international library.

* Font-sizes for widgets on GedaToolbars can now be specified in RC files.

* The toolbar layout in gschem has been improved to better support labels.

* The visibility of toolbar tooltips can now be control using RC keywords.

* The Slot Editor Dialog now incorporates a spinner control to adjust the
  slot number of slotted components. The range is restricted to valid slots
  for the selected component.

* New bitmaps have been added;

      * geda-deslect.png
      * geda-find-attribute
      * geda-properties.png
      * geda-trash-can.png

* libgedathon no longer over-writes existing gafrc when the declare_local_sym
  feature is used.

* libgedathon now supports internationalization (i18n).

Notable changes in gEDA/gaf gEDA 2.1.1.0 - 01/01/17 Rampage Edition

* Improvements were made to both the build system and source code to ensure
  gEDA-GAF builds more reliably on 64 bit platforms. Testing was performed
  using Ubuntu 12.04 and Ubuntu 14.04 64 bit host systems.

* The libgeda API has been completely redone in order to present a library API
  that is not only for in-suite applications.

* Support for internationalization (i18n) has been significantly expanded and
  now includes command-line utilities.

* Extensive revisions have been made to improve performance of gEDA application.
  This is particularly noticeable in gschem.

* A new library, libgedacolor, now unifies color management for the gEDA suite.
  Color-maps are still configured using scheme files, but the use of "friendly-
  color-maps"  is no longer applicable; the new color maps for libgedacolor are
  already "friendly". The library was added to provide a unified API for gEDA
  applications and should be transparent to most users. User who are using
  custom color-maps will need to modify their custom map files.

* A new auto-panning feature has been added to gschem. The auto-pan is only in
  effect while drawing nets and buses and can enabled or disabled through menu
  options.

* The Auto-Number dialog in gschem no longer requires the use of wild-card
  characters in the scope text. This is a new advanced feature that allows
  automatic re-numbering of duplicated reference designators.

* The path for logs file can now be set in initialization files, including
  local gafrc files. This allows logs to be organized on a per project basis.
  The location for logs can also be set using a new GEDALOGS environment variable.

* A new libgeda keyword "log-time" can be use to enable or disable prefixing
  log entries with the time of day.

* The Multi-Attribute Editor Dialog entry now has attribute "word" completion
  identical to the Single-Attribute Editor Dialog.

* The list of valid attribute used by gsymcheck when checking symbol files is
  now users configurable. The list can be set in system-gsymcheckrc.

* The list of known devices recognized by gsymcheck when checking the device
  entries in symbol files is now users configurable. The list can be set in
  system-gsymcheckrc.

* Symbol files can now be renamed using the Component Dialog.

* The number of user defined colors was increased by 6 to a total of
  10 user definable colors in each color map.

* The main menu system has been modified to automatically build a basic
  fall-back menu when errors are discovered in the menu RC file.

* The height of the status bar in Gschem can now be set in the gschem-gtkrc

* HotKey dialog now includes actions icon and a column to display the actual
  binding string. User can enable or disable viewing the action column with
  the setting being retained between session.

* The Slot Editor Dialog now displays the total number of slots per component.

* New features have been added to the Page Select Dialog:

      Accept drag-and-drop of schematic or symbol files.
      Opened pages can be re-ordered either up or down.
      Auto scrolls to selection.
      Shows long filename as tooltip when showing short names.
      Optionally adjust height of dialog automatically.

* Exchange image file now retains the user preference to update all selected
  images between session. Previously the user preference was only retained for
  current session.

* Additional preference variables have been incorporated in new the configuration
  system allow retention of changes using the Preferences dialog:

      continue-component-place
      embed-components
      enforce-hierarchy
      force-boundingbox
      keyboardpan-gain
      netconn-rubberband
      select-slack-pixels
      snap-size
      sort-component-library
      object-clipping

  Uses of these keywords in rc files is still permissible and will over-ride the
  preference setting.

* A new gschem command line options has been provided to specify a title-block at
  startup.

* A new gschem command line options has been provided to startup minimized.

* The command line option to add a directory to the Guile path now has a long option.
  The "-L" option has been replaced with "-g" or "--guile".

* gsymcheck now supports the standard command-line version options.

* Programs in the geda-gaf suite now support quite mode in conjunction with the version
  option to assist with extraction of dotted version information.

* New Scheme API have been added:

      gschem-bmp-image
      gschem-jpeg-image
      gschem-tiff-image

* The maximum zoom magnification in gschem has been increased slightly.


libgedathon/scripts/gsym-set-font-size.py
#! /usr/bin/env python

# Set Visibility of Attributes (For now just to Defaults)

import os, sys, glob
from geda import geda
from geda.constants import *

#---------------------------------------------------------------------------
Version="0.2.0"
#---------------------------------------------------------------------------
VerboseMode=False
#---------------------------------------------------------------------------

#  This is the help string.
Usage =\
"""
gsym-set-font-size -- A gEDA-gaf Symbol File Utility
                      to set the font size of Text attributes in a symbol file
                      to default values or as otherwise specified

Usage: gsym-set-font-size [Options] -i <inputfile> [[-o] <outputfile> ]
"""

Help =\
"""
Options:

  -r, --recursive -- Process all symbol files in the current and all subordinated directories.
  -v, --verbose   -- Verbose mode.  Used in both archive and extract mode.
                     Spews lots of info about what the program is doing.

  The following options can be used to modify the default behavior of gsym-set-font-size
  The following list of attribute flags correspond to attributes set to 10 points by default,
  if these options are used the corresponding attributes will not be modified:

  -D, --device
  -R, --refdes
  -A, --author
  -U, --numslots
  -E, --description
  -C, --comment
  -N, --net
  -M, --netname
  -V, --value
  -L, --pinlabel

  The following list of attribute flags correspond to attributes set to 8 points by default,
  if these options are used the corresponding attributes will not be modified:

  -q, --pinseq
  -p, --pinnumber
  -t, --pintype

  The following list of attributes are set to 10 points, and no option is provided to
  change this behavior:

  "documentation", "symversion", "dist-license", "use-license", "file", "footprint",
  "footprints", "graphical", "pins", "slot", "slotdef", "spice-type", "electtype",
  "mechtype", "model-name"

Example: Set the font size of all attributes to default values in all symbol files
         in the current directory and in all subdirectories:

         gsym-set-font-size.py -r

Copyright (C) 2014-2015 by Wiley Edward Hill.  Released under GPL Version 2.

"""

#-------------------------------- Classes ----------------------------------
class ProgramParameters:
    """
    This class holds info about the environment and cmd line args passed to the
    program.  It has only one method:  the constructor, which gets the args
    and fills out the public vars.  The public vars are:
    """
    def __init__(self):

        global Version

        """
        Constructor: parse through cmd line args and fill out vars. The values
        here are the defaults
        """
        self.RecursiveMode   = False
        self.VerboseMode     = False
        self.InputFiles      = []
        self.UserDir         = os.path.abspath(os.getcwd())

        self.fix_device      = True
        self.fix_refdes      = True
        self.fix_author      = True
        self.fix_numslots    = True
        self.fix_description = True
        self.fix_comment     = True
        self.fix_net         = True
        self.fix_netname     = True
        self.fix_value       = True
        self.fix_pinlabel    = True

        self.fix_pinseq      = True
        self.fix_pinnumber   = True
        self.fix_pintype     = True

        if len(sys.argv) > 1:
            for arg in sys.argv[1:]:       # Skip OurSelf
                if arg in ('-V', '--version'):
                    print Version
                    sys.exit(0)
                if arg in ('-h', '--help'):
                    print Usage
                    print Help
                    sys.exit(0)
                if arg in ('-v', '--verbose'):
                    self.VerboseMode = True
                    continue
                if arg in ('-r', '--recursive'):
                    self.RecursiveMode = True
                    continue

                # The set to 10 points group
                if arg in ('-D', '--device'):
                    self.fix_device = False
                    continue
                if arg in ('-R', '--refdes'):
                    self.fix_refdes = False
                    continue
                if arg in ('-A', '--author'):
                    self.fix_author = False
                    continue
                if arg in ('-U', '--numslots'):
                    self.fix_numslots = False
                    continue
                if arg in ('-E', '--description'):
                    self.fix_description = False
                    continue
                if arg in ('-C', '--comment'):
                    self.fix_comment = False
                    continue
                if arg in ('-N', '--net'):
                    self.fix_net = False
                    continue
                if arg in ('-M', '--netname'):
                    self.fix_netname = False
                    continue
                if arg in ('-V', '--value'):
                    self.fix_value = False
                    continue
                if arg in ('-L', '--pinlabel'):
                    self.fix_pinlabel  = False
                    continue

                # The set to 8 points group

                if arg in ('-q', '--pinseq'):
                    self.fix_symversion = False
                    continue
                if arg in ('-p', '--pinnumber'):
                    self.fix_pinnumber = False
                    continue
                if arg in ('-t', '--pintype'):
                    self.fix_pintype = False
                    continue

                self.InputFiles.append(arg)


#---------------------------------------------------------------------------
# The ALWAYS turn OFF goup
def AlwaysSet10(Options, Text):

    list = [ "documentation", "symversion", "dist-license", "use-license", "file",
             "footprint", "footprints", "graphical", "pins", "slot", "slotdef",
             "spice-type", "electtype", "mechtype", "model-name"]

    status = 0
    string = Text.name()
    for name in list:
        if name == string:
            status = 1
            if not Text.size == 10:
                Text.size = 10
                status = 2
            break

    return status

#---------------------------------------------------------------------------

# The Set to 8 points group
def SetTextSize8 (Options, Text):
    if Text.name() == "pinseq":
        status = 1
        if Options.fix_pinseq and not Text.size == 8:
            status = 2
    elif Text.name() == "pinnumber":
        status = 1
        if Options.fix_pinnumber and not Text.size == 8:
            status = 2
    elif Text.name() == "pintype":
        status = 1
        if Options.fix_pintype and not Text.size == 8:
            status = 2
    else:
        status = 0

    if status == 2:
        Text.size = 8

    return status

#---------------------------------------------------------------------------
# The Set to 10 points group
def SetTextSize10 (Options, Text):
    if Text.name() == "device":
        status = 1
        if Options.fix_device and not Text.size == 10:
            status = 2
    elif Text.name() == "refdes":
        status = 1
        if Options.fix_refdes and not Text.size == 10:
            status = 2
    elif Text.name() == "author":
        status = 1
        if Options.fix_author and not Text.size == 10:
            status = 2
    elif Text.name() == "numslots":
        status = 1
        if Options.fix_numslots and not Text.size == 10:
            status = 2
    elif Text.name() == "description":
        status = 1
        if Options.fix_description and not Text.size == 10:
            status = 2
    elif Text.name() == "comment":
        status = 1
        if Options.fix_comment and not Text.size == 10:
            status = 2
    elif Text.name() == "net":
        status = 1
        if Options.fix_net and not Text.size == 10:
            status = 2
    elif Text.name() == "netname":
        status = 1
        if Options.fix_netname and not Text.size == 10:
            status = 2
    elif Text.name() == "value":
        status = 1
        if Options.fix_value and not Text.size == 10:
            status = 2
    elif Text.name() == "pinlabel":
        status = 1
        if Options.fix_pinlabel and not Text.size == 10:
            status = 2
    else:
        status = 0

    if status == 2:
        Text.size = 10

    return status

#---------------------------------------------------------------------------

def AnalyzeText(Options, Text):
    modified = False
    string = Text.string
    index = string.find("=")
    if index > 0:
        modified = SetTextSize10(Options, Text)
        if not modified:
            modified = SetTextSize8(Options, Text)
            if not modified:
                modified = AlwaysSet10(Options, Text)
        modified = modified - 1

    return modified

#---------------------------------------------------------------------------

def ProcessSymbol(Options, File):

    symbol = geda.open_page(File)

    if Options.VerboseMode:
        print "Processing: " + symbol.filename

    objects  = geda.get_objects(symbol)
    modified = 0
    """
      Loop thru all objects and get list of detached text attribute then loop
      again and skip if in list, Sub object (attrinutes) are to be handled by
      the objects. This is so that detached text gets added to the symbol sym
      groups.
    """
    for capsule in objects:
        object = geda.get_object(capsule)          # Retrieve object from GedaCapsule
        if capsule.type == OBJ_TEXT:
            result = AnalyzeText(Options, object)
            if result > 0:
                modified = modified + 1

    if modified:
        if Options.VerboseMode:
            print "Made " + str(modified) + " modifications to " + symbol.filename
        symbol.save() # Write the symbol to storage

    # Close the file
    symbol.close()

#---------------------------------------------------------------------------

def RecursiveGlob(path, *exts):
	""" Glob recursively a directory and all subdirectories for multiple file extensions
    Note: Glob is case-insensitive, i. e. for '\*.jpg' you will get files ending
    with .jpg and .JPG

    Parameters
    ----------
    path : str
        A directory name
    exts : tuple
        File extensions to glob for

    Returns
    -------
    files : list
        list of files matching extensions in exts in path and subfolders
	"""
	dirs = [a[0] for a in os.walk(path)]
	f_filter = [d+e for d in dirs for e in exts]
	return [f for files in [glob.iglob(files) for files in f_filter] for f in files]

#---------------------------------------------------------------------------
#  Main prog begins here

def main(argv):

	foundOurSelf = False

	Options = ProgramParameters()  #  Creates Options object for program varibles.

	InputFiles = Options.InputFiles

	files = []
	if len(InputFiles) == 0:
		# Get list of file OurSelf
		foundOurSelf = True
		if Options.RecursiveMode:
			files = RecursiveGlob(os.getcwd(), '/*.sym')
		else:
			files = glob.glob('*.sym')

	if len(files) > 0:
		for file in files:
			InputFiles.append(file)

	if len(InputFiles) == 0:
		print 'No symbols found or specifed, try using --help'
	elif not foundOurSelf :
		for file in InputFiles:
			if not os.path.isfile(file):
				print "Bad file name or option: " + file
				print Usage
			else:
				ProcessSymbol(Options, file)

	else:
		for symbol in InputFiles:
			ProcessSymbol(Options, symbol)

if __name__ == "__main__":
   main(sys.argv[1:])

# Done!

